include/mqtt/endpoint.hpp

```diff
@@ -1779,39 +1797,41 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
             contents_buf,
             pubopts,
             force_move(props),
-            force_move(func),
-            std::make_pair(force_move(sp_topic_name), force_move(sp_contents))
+            [topic_life   = force_move(sp_topic_name),
+             content_life = force_move(sp_contents),
+             func         = force_move(func)]
+            (error_code ec) {
+                if(func) func(ec);
+            }
         );
     }
 
     /**
      * @brief Publish with a manual set packet identifier
      * @param packet_id
      *        packet identifier. It should be acquired by acquire_unique_packet_id, or register_packet_id.
-     *        The ownership of  the packet_id moves to the library.
+     *        The ownership of the packet_id moves to the library.
      *        If qos == qos::at_most_once, packet_id must be 0. But not checked in release mode due to performance.
      * @param topic_name
      *        A topic name to publish
      * @param contents
      *        The contents to publish
-     * @param life_keeper
-     *        An object that stays alive (but is moved with force_move()) until the async operation is finished.
      * @param pubopts
      *        qos, retain flag, and dup flag.
      * @param func
      *        functor object who's operator() will be called when the async operation completes.
+     *        This object should hold the lifetime of the buffers for topic_name and contents.
      */
     void async_publish(
         packet_id_t packet_id,
         as::const_buffer topic_name,
         as::const_buffer contents,
-        any life_keeper,
-        publish_options pubopts = {},
-        async_handler_t func = async_handler_t()
+        publish_options pubopts,
```

Comment:
Why do you remove life_keeper ?

Reply:
Because the way that boost.asio completion tokens are intended to work is that the lifetime is held by the completion token, which is also the async_handler.

Reply:
I think that the lifetime of topic_name and contents are different from async_handler_t.

For example, QoS1 publish,

1. topic_name and contents lifetime are held by func (lambda captured).
2. The topic_name and contents lifetime should be copied to store_.
3. async_write is requested, and func is copied (or moved) to asio.
4. async_write completion token is called and func's lifetime is over (also captured variables)
5. disconnect and connect again.
6. send from store but topic_name and contents are destroyed at the step4.

Am I missing something?

Reply:
Notice that inside of the send_async_publish function, I take a copy of the async_handler_t, to preserve the lifetime of the topic and contents.

Reply:
Ok, I understand.
I think that there are three API design candidates.

1. Remove all life_keeper (your PR).
On publish implementation, as you mentioned, we need to store func to keep lifetime for captured variables. I think that user expects after func is called, the capture object is destroyed. But the actual behavior is different.
Consider this example:
```cpp
publish(... [sp_topic_name, sp_contents, large_object_for_user_callback](...) {
    // use large_object_for_user_callback here
});
```
It is unexpected.

2. Keep all life_keeper (the current master)
It is very clear meaning. life_keeper is only used to keep the lifetime. async_handler_t is only used to notify async operation is finished. The disadvantage is the number of parameters. But it is not a big problem.

3. Keep life_keeper only publish
It is inconsistent and difficult to understand.

So I prefer to the approach 2. I think that it doesn't avoid the way to Boost.Asio CompletionToken.

Reply:
As far as I am aware, boost.asio completion tokens do not provide the guarantee that the completion token will be deleted soon after being called.

I also can't find any documentation that says they will only be called once, but it seems to me like they would have to be in order for the coroutine support to work.

I am against having the "life_keeper" variable at all. I think that the storage for at_least_once / exactly_once should not be handled by endpoint.hpp directly, and instead should be virtual function calls so that the end user can handle them. The end user would be able to provide the code needed to handle lifetime the way they want to, without surprises.

However, I do understand your point about end user confusion.

I would strongly prefer option 1, but will accept option 3.

If I update the PR to be option 3, is that acceptable?

Reply:
> As far as I am aware, boost.asio completion tokens do not provide the guarantee that the completion token will be deleted soon after being called.
> 
> I also can't find any documentation that says they will only be called once, but it seems to me like they would have to be in order for the coroutine support to work.

Even if it is not documented, if user pass a callback function as CompletionToken, it should be called only once and immediately deleted. I observed Boost Asio behavior and it works as I expected. As you pointed out passing coroutine_t scenario is different. I agree with that. I guess that the current document of ConpletionToken needs to be abstract. But callback function scenario works as I expected. It is now undocumented so it works by chance, so far. I think that ConpletionToken detail should be documented in the future.

I am strongly against 1. 

> If I update the PR to be option 3, is that acceptable?

It is acceptable because it is the same as the last release v7.0.1.  
https://github.com/redboltz/mqtt_cpp/blob/v7.0.1/include/mqtt/endpoint.hpp#L2910

Let's go approach 3!

Reply:
> Let's go approach 3!

Great!

Will update in a day or two.

Reply:
Updated.

Please also review async_pubrel and async_pubcomp, since they are related to async_publish.